csrc/ops/composite.cpp

```diff
@@ -53,15 +53,22 @@ TensorView* dropout_backward(TensorView* dy, TensorView* mask, Val* scale) {
   return dx;
 }
 
-TensorView* matmul(TensorView* a, TensorView* b) {
+namespace {
+
+// This function will be used by both matmul and linear.
+// Matmul will add a castOp to the output of this function
+// whereas linear will skip the cast and add the bias.
+TensorView* matmulImpl(TensorView* a, TensorView* b) {
```

Comment:
Suggestion: instead of an "impl" function, what if we accept an additional optional `DataType output_dtype` arg that defaults to `DataType::Null`. If that default is given then we replace it with the dtype of `a`. Then we can use `SimplifyingIrBuilder::maybeCastExpr` to cast to that dtype. That way we could still use the familiar `matmul()` interface to create matmuls with single-precision outputs when we want to by providing `/*output_dtype=*/DataType::Float`, but the default would be to downcast.

Reply:
Thanks - I switched to using `maybeCastOp`, but I think I'd like to keep the Impl function and have a very thin matmul function. I'd prefer not to have that flag directly in matmul itself (thought I don't feel very strongly about it).

Reply:
My reasoning was so that we could actually create a matmul with single-precision output from outside this compilation unit. As is currently in this PR, that functionality is hidden even in C++ is it not?

Reply:
We don't need to expose this function in the python frontend, where I agree that we should keep the surface small. But on the C++ side it seems useful to be able to access single-precision output `matmul`, for example when creating matmuls with epilogues that don't fit the `linear` template.

Reply:
Thanks for pointing that out. To enable using this from the C++ side without the cast I overloaded the matmul. Currently the python binding implementation is a bit messy which made calling the same matmul function with default args a bit tricky.